### PR TITLE
APPS/IO-DEMO: Backport report performance fixes from master

### PR DESCRIPTION
**Pull request by dmitrygx**
1. Fix reporting performance.
2. Fix fairness of sending IO data.

there were some merge conflict with master:
- the 2nd commit: minor conflict related to AM code where `get_num_uncompleted()` is defined
- the 3rd commit: don't pass data_size for AM code.

### CODE REVIEW

File: test/apps/iodemo/io_demo.cc [hunk 1]
```diff
@@ -1173,25 +1198,50 @@ class DemoClient : public P2pDemoCommon {
...
         return get_num_uncompleted(_server_info[server_index]);
     }
 
+    static long get_num_completed_round(const server_info_t& server_info,
+                                        io_op_t op)
+    {
+        return server_info.num_completed[op] - server_info.prev_completed[op];
+    }
+
+    static long get_num_posted_round(const server_info_t& server_info,
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: where is it used?

File: test/apps/iodemo/io_demo.cc [hunk 2]
```diff
@@ -1544,58 +1592,91 @@ class DemoClient : public P2pDemoCommon {
...
 
-    void report_performance(long num_iters, double elapsed,
-                            size_t total_bytes[IO_OP_MAX]) {
+    struct io_op_perf_info_t {
+        long min; // Minimum number of completed operations on some server
+        long max; // Maximum number of completed operations on some server
+        size_t min_index; // Server index with the smallest number of completed
+                          // operations
+        long total; // Total number of completed operations
+        size_t total_bytes; // Total number of bytes of completed operations
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: align on column

File: test/apps/iodemo/io_demo.cc [hunk 3]
```diff
@@ -1544,58 +1592,91 @@ class DemoClient : public P2pDemoCommon {
...
+            size_t total_bytes_completed = 0;
+            for (int op = 0; op <= IO_OP_MAX; ++op) {
+                size_t bytes_completed;
+                long delta_completed;
+                if (op != IO_OP_MAX) {
+                    assert(server_info.bytes_sent[op] ==
+                                   server_info.bytes_completed[op]);
+                    bytes_completed = server_info.bytes_completed[op];
+                    delta_completed =
+                            get_num_completed_round(server_info,
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: seems get_num_completed_round() used only once, can be removed?

File: test/apps/iodemo/io_demo.cc (hunk 1 above)
Comment (dmitrygx) [reply test/apps/iodemo/io_demo.cc]: removed

File: test/apps/iodemo/io_demo.cc (hunk 2 above)
Comment (dmitrygx) [reply test/apps/iodemo/io_demo.cc]: done

File: test/apps/iodemo/io_demo.cc (hunk 3 above)
Comment (dmitrygx) [reply test/apps/iodemo/io_demo.cc]: done

**Review by yosefe — APPROVED** [review]
can u pls squash and port fixes to master?

**Pull request merged**